packages/core/database/src/index.ts

```diff
@@ -92,16 +107,37 @@ class Database {
     this.lifecycles = createLifecyclesProvider(this);
 
     this.entityManager = createEntityManager(this);
-
-    this.logger = config.logger ?? console;
   }
 
   async init({ models }: { models: Model[] }) {
+    if (typeof this.config.connection.connection === 'function') {
+      // for function connections, we must configure the dialect in init
+      // Note: due to a bug in Knex, sqlite will warn about a missing filename for function connections because
+      // it checks in the constructor rather than on connect
+      const conn = await this.config.connection.connection();
```

Comment:
It would probably be better from a performance perspective to only run the token acquisition when needed. I suspect (though am not sure) that getting it twice won't invalidate the first one, but it's still a startup and code cost.

Given that all you're doing with the connection from that point seems to be getting a string value, why not just get the string from the client string from the config object rather than the connection?

I don't see it as inherently more reliable to get it off the connection function -- it's coming from the config value anyway. Unless I'm reading the code wrong.

Reply:
it's a bit of a catch-22 that requires refactoring of our dialect code to get right, but I was worried about breaking something. Basically, dialect needs access to the connection config before we can use the knex connection. But we don't have access to the connection config object until the connection is used.

I might be able to do some safe refactoring to get it a bit closer, but I don't think we'll be able to get it perfect until we make some breaking changes. And even with breaking changes, it may be impossible in cases like sqlite where we want to verify a filename exists before we create the knex connection, but I think it would be fine there to just disable the check and have it be the responsibility of the user to ensure the file exists if they're using a connection function.

Reply:
Maybe it's worth lifting the config object up the initialisation path? Anything lower down the path would still be able to access it, and you'd be able to make use of it much earlier. 

Obviously don't know the code well, but that seems like something that could be done without breakage?

Just feels (from the outside, without full context) like you're taking on complexity to avoid a larger piece of work which, let me be absolutely clear about, I have 100% done in my life, so I completely get it, but I have also regretted it more times than I have not.

Reply:
The main problem is that we can't, until we are able to make a breaking change (the one I missed the deadline on making for v5) just due to how everything that was sync has to remain sync, and we need to be able to modify the user knex config before passing it to knex.

However, I just pushed an update that should solve everything. I tested it out locally with expiring connections and it seems to work great. I will generate a new experimental and post it in a few minutes.

Reply:
In v6 we will just declare "you cannot rely on connection.connection to be an object and you must wait until db init before accessing it" and update our code internally as well, and can remove any added complexity here